File: internal/features/scratch/log.go
package scratch

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"hdd/internal/logging"
	"hdd/internal/model"
)

func AppendScratchRoundLog(logDir string, email string, result model.ScratchRoundResult, totalCost float64, totalReward float64) error {
	if logDir == "" {
		return nil
	}
	return appendScratchLogLine(logDir, email, formatScratchRoundLogLine(result, totalCost, totalReward))
}

func appendScratchLogLine(logDir string, email string, line string) error {
	path := logging.LogFilePath(logDir, email)
	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
		return err
	}
	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0600)
	if err != nil {
		return err
	}
	defer file.Close()
	if _, err := file.WriteString(line); err != nil {
		return err
	}
	return file.Sync()
}

func formatScratchRoundLogLine(result model.ScratchRoundResult, totalCost float64, totalReward float64) string {
	balanceSuffix := roundBalanceSuffix(result)
	if result.PlayErr != nil {
		return fmt.Sprintf("===== 第 %d 轮结束，用时 %s =====\n这一轮开局失败了：%v\n累计情况：到现在一共花了 %.2f，一共中了 %.2f，净收益 %.2f%s。\n",
			result.Round,
			result.Duration.Round(time.Millisecond),
			result.PlayErr,
			totalCost,
			totalReward,
			totalReward-totalCost,
			balanceSuffix,
		)
	}
	line := fmt.Sprintf("===== 第 %d 轮结束，用时 %s =====\n", result.Round, result.Duration.Round(time.Millisecond))
	if result.PlayResp != nil {
		line += fmt.Sprintf("这一轮已经开局：玩法是%s，对局编号是 %d，目前状态是%s，这一局花了 %.2f，当前余额 %.8f。\n",
			gameTypeLabel(result.PlayResp.GameType),
			result.PlayResp.PlayID,
			playStatusLabel(result.PlayResp.Status),
			result.PlayResp.CostAmount,
			result.PlayResp.Balance,
		)
	}
	if result.PlayHistoryErr != nil {
		line += fmt.Sprintf("开局记录同步失败：%v\n", result.PlayHistoryErr)
		line += fmt.Sprintf("累计情况：到现在一共花了 %.2f，一共中了 %.2f，净收益 %.2f%s。\n", totalCost, totalReward, totalReward-totalCost, balanceSuffix)
		return line
	}
	if result.RevealErr != nil {
		line += fmt.Sprintf("这一轮开奖失败了：%v\n", result.RevealErr)
		line += fmt.Sprintf("累计情况：到现在一共花了 %.2f，一共中了 %.2f，净收益 %.2f%s。\n", totalCost, totalReward, totalReward-totalCost, balanceSuffix)
		return line
	}
	if result.RevealResp != nil {
		line += fmt.Sprintf("这一轮已经开奖：玩法是%s，对局编号是 %d，结果是%s，奖金 %.2f，净收益 %.2f，当前余额 %.8f。\n",
			gameTypeLabel(result.RevealResp.GameType),
			result.RevealResp.PlayID,
			roundOutcomeLabel(model.ScratchRoundResult{RevealResp: result.RevealResp}),
			result.RevealResp.RewardAmount,
			result.RevealResp.NetAmount,
			result.RevealResp.Balance,
		)
	}
	if result.RevealHistoryErr != nil {
		line += fmt.Sprintf("开奖记录同步失败：%v\n", result.RevealHistoryErr)
	}
	line += fmt.Sprintf("累计情况：到现在一共花了 %.2f，一共中了 %.2f，净收益 %.2f%s。\n", totalCost, totalReward, totalReward-totalCost, balanceSuffix)
	return line
}

File: src/workflows/scratch/log.rs
use std::io;
use std::path::Path;

use crate::model::{
    SCRATCH_GAME_TYPE_ICON_MATCH, SCRATCH_GAME_TYPE_LUCKY_NUMBERS, SCRATCH_GAME_TYPE_PROGRESS_RUN,
    SCRATCH_GAME_TYPE_THREE_KIND, SCRATCH_GAME_TYPE_TREASURE_CHEST, ScratchRoundResult,
};
use crate::ui;
use crate::workflows::common::{
    append_account_log_line, format_amount, format_duration_ms as common_format_duration_ms,
};

pub(super) fn append_scratch_round_log(
    log_dir: &Path,
    email: &str,
    result: &ScratchRoundResult,
    total_cost: f64,
    total_reward: f64,
) -> io::Result<()> {
    append_account_log_line(
        log_dir,
        email,
        &format_scratch_round_log_line(result, total_cost, total_reward),
    )
}

fn format_scratch_round_log_line(
    result: &ScratchRoundResult,
    total_cost: f64,
    total_reward: f64,
) -> String {
    let mut lines = Vec::new();
    lines.push(format!(
        "===== 第 {} 轮结束，用时 {} =====",
        result.round,
        format_duration_ms(result.duration_ms)
    ));

    if !result.play_error_message.is_empty() {
        lines.push(format!("这一轮开局失败了：{}", result.play_error_message));
        lines.push(format_stats_line(
            total_cost,
            total_reward,
            round_balance_suffix(result),
        ));
        return lines.join("\n") + "\n";
    }

    if let Some(play_resp) = &result.play_resp {
        lines.push(format!(
            "这一轮已经开局：玩法是{}，对局编号是 {}，目前状态是{}，这一局花了 {}，当前余额 {}。",
            game_type_label(&play_resp.game_type),
            play_resp.play_id,
            play_status_label(&play_resp.status),
            format_amount(play_resp.cost_amount),
            format_amount(play_resp.balance),
        ));
    } else if let Some(play_history_item) = &result.play_history_item {
        lines.push(format!(
            "这一轮先补开之前没处理完的旧对局：对局编号 {}，原来状态是{}，这一局花了 {}。",
            play_history_item.id,
            play_status_label(&play_history_item.status),
            format_amount(play_history_item.cost_amount.unwrap_or(0.0)),
        ));
    }

    if !result.play_history_error_message.is_empty() {
        lines.push(format!(
            "开局记录同步失败：{}",
            result.play_history_error_message
        ));
        lines.push(format_stats_line(
            total_cost,
            total_reward,
            round_balance_suffix(result),
        ));
        return lines.join("\n") + "\n";
    }
    if !result.reveal_error_message.is_empty() {
        lines.push(format!("这一轮开奖失败了：{}", result.reveal_error_message));
        lines.push(format_stats_line(
            total_cost,
            total_reward,
            round_balance_suffix(result),
        ));
        return lines.join("\n") + "\n";
    }
    if let Some(reveal_resp) = &result.reveal_resp {
        lines.push(format!(
            "这一轮已经开奖：玩法是{}，对局编号是 {}，结果是{}，奖金 {}，净收益 {}，当前余额 {}。",
            game_type_label(&reveal_resp.game_type),
            reveal_resp.play_id,
            round_outcome_label(result),
            format_amount(reveal_resp.reward_amount),
            format_amount(reveal_resp.net_amount),
            format_amount(reveal_resp.balance),
        ));
    }
    if !result.reveal_history_error_message.is_empty() {
        lines.push(format!(
            "开奖记录同步失败：{}",
            result.reveal_history_error_message
        ));
    }
    lines.push(format_stats_line(
        total_cost,
        total_reward,
        round_balance_suffix(result),
    ));
    lines.join("\n") + "\n"
}

pub(super) fn log_round_result(
    log: &ui::TaskLog,
    result: &ScratchRoundResult,
    total_cost: f64,
    total_reward: f64,
) {
    log.line(format_scratch_round_log_line(
        result,
        total_cost,
        total_reward,
    ));
}

fn format_stats_line(total_cost: f64, total_reward: f64, balance_suffix: String) -> String {
    format!(
        "累计情况：到现在一共花了 {}，一共中了 {}，净收益 {}{}。",
        format_amount(total_cost),
        format_amount(total_reward),
        format_amount(total_reward - total_cost),
        balance_suffix
    )
}

fn round_balance_suffix(result: &ScratchRoundResult) -> String {
    match round_balance(result) {
        Some(balance) => format!("，当前余额 {}", format_amount(balance)),
        None => String::new(),
    }
}

fn round_balance(result: &ScratchRoundResult) -> Option<f64> {
    if let Some(reveal_resp) = &result.reveal_resp {
        return Some(reveal_resp.balance);
    }
    if let Some(play_resp) = &result.play_resp {
        return Some(play_resp.balance);
    }
    None
}

fn round_reward(result: &ScratchRoundResult) -> f64 {
    if let Some(history) = &result.reveal_history_item {
        return history.reward_amount.unwrap_or(0.0);
    }
    if let Some(reveal) = &result.reveal_resp {
        return reveal.reward_amount;
    }
    0.0
}

fn round_outcome_label(result: &ScratchRoundResult) -> &'static str {
    if result.reveal_resp.is_none() && result.reveal_history_item.is_none() {
        return "本局未完成";
    }
    if round_reward(result) > 0.0 {
        return "中奖了";
    }
    "未中奖"
}

fn game_type_label(game_type: &str) -> &str {
    match game_type.trim() {
        SCRATCH_GAME_TYPE_LUCKY_NUMBERS => "幸运数字",
        SCRATCH_GAME_TYPE_THREE_KIND => "三连相同",
        SCRATCH_GAME_TYPE_ICON_MATCH => "图标配对",
        SCRATCH_GAME_TYPE_TREASURE_CHEST => "宝箱开奖",
        SCRATCH_GAME_TYPE_PROGRESS_RUN => "进度冲刺",
        _ if game_type.trim().is_empty() => "未知玩法",
        _ => game_type.trim(),
    }
}

fn play_status_label(status: &str) -> &str {
    match status.trim().to_ascii_lowercase().as_str() {
        "pending" => "等待开奖",
        "played" => "已开局",
        "revealed" | "done" | "granted" => "已开奖",
        _ if status.trim().is_empty() => "状态未知",
        _ => status.trim(),
    }
}

fn format_duration_ms(duration_ms: i64) -> String {
    common_format_duration_ms(duration_ms)
}
